### ISSUE TITLE
[REVIEW]: Blimpy: Breakthrough Listen I/O Methods for Python

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1554]
**Submitting author:** @telegraphic (<a href="http://orcid.org/0000-0003-2783-1608">Danny Price</a>)
**Repository:** <a href="https://github.com/UCBerkeleySETI/blimpy" target ="_blank">https://github.com/UCBerkeleySETI/blimpy</a>
**Version:** 1.4.1
**Editor:** @xuanxu
**Reviewer:** @garrettj403, @zhampel
**Archive:** <a href="https://doi.org/10.5281/zenodo.3472084" target="_blank">10.5281/zenodo.3472084</a>

## Status

[![status](http://joss.theoj.org/papers/e58ef21f0a924041bf9438fd75f8aed0/status.svg)](http://joss.theoj.org/papers/e58ef21f0a924041bf9438fd75f8aed0)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/e58ef21f0a924041bf9438fd75f8aed0"><img src="http://joss.theoj.org/papers/e58ef21f0a924041bf9438fd75f8aed0/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/e58ef21f0a924041bf9438fd75f8aed0/status.svg)](http://joss.theoj.org/papers/e58ef21f0a924041bf9438fd75f8aed0)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/UCBerkeleySETI/blimpy" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@garrettj403 & @zhampel, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @xuanxu know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/UCBerkeleySETI/blimpy">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** 1.4.1
- [x] **Authorship:** Has the submitting author (@telegraphic) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/UCBerkeleySETI/blimpy">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** 1.4.1
- [x] **Authorship:** Has the submitting author (@telegraphic) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @garrettj403, @zhampel it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf)

**Comment by zhampel** [comment]
@xuanxu, @telegraphic, there are a couple of GitHub users (beyond the listed authors) who have contributed to the repo. Specifically FX196 & marksbrt. 

@xuanxu are there clear guidelines on defining contributing authors. I just want to check prior to checking some boxes.

**Comment by xuanxu** [comment]
@zhampel authorship is a complex topic, here are some guidelines: https://joss.readthedocs.io/en/latest/review_criteria.html#authorship

*the authors themselves are responsible for deciding which contributions are sufficient for co-authorship. Your job as a reviewer is to check that the list of authors appears reasonable, and if it’s not obviously complete/correct, to raise this as a question during the review.*

**Comment by zhampel** [comment]
Yeah, it's that last phrase why I ask. I'll check off the box for now, unless the authors want to comment otherwise.

**Comment by garrettj403** [comment]
Hi, sorry for the delay -- I've been really busy for the last two weeks, but I have a lot more time now. I started by taking a look at the documentation, and I'll try out the software over the next few days.

JOSS paper
----------
- I have the same question as @zhampel about authorship. In the ``setup.py`` file, there are 4 authors listed (Danny Price, Emilio Enriquez, Griffin Foster, Greg Hellbourg), and on GitHub, there are another two authors who seem to have made substantial commits (Yuhong Chen and Mark Siebert). 
- The paper is still very short. I think @kyleniemeyer  brought this up in the pre-review, but it hasn't been addressed so far. It would be nice if you could elaborate a bit more (e.g., what exactly you're looking for in the data, what is a "waterfall" file, etc.).
- The references don't display correctly in the [PDF proof](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf). This might be an issue with the formatting of the ``paper.bib`` file. 
- The ``Croft:2016`` reference has a DOI listed online that is missing from the ``paper.bib`` file (10.13140/RG.2.2.10561.20326). Also, this is not a published ``article``. It should be listed as a ``techreport`` or ``unpublished``, and include a URL to the paper. ``researchgate.net`` also shouldn't be listed as the journal name.
- I wasn't able to find a software archive for ``Blimpy``.
- There are several typos in ``paper.md``:
   - avaiable -> available
   - guppi -> GUPPI
   - astromers -> astronomers
   - endeavour -> endeavor

Documentation
-------------
- The [project API](https://blimpy.readthedocs.io/en/latest/modules.html) is very difficult to read/navigate. Most functions and methods are missing argument descriptions. 
- The examples are very limited. The only examples that I could find are in ``README.md``, but they only really show how to import data.
- In ``README.md``, it says "This readme is far from complete. If you have any request/questions please lets us know!". The REAME should be expanded.
- Overall, I'm finding it very difficult to begin using the software with the current level of documentation.

**Comment by garrettj403** [comment]
Functionality
-------------
- You should mention in the documentation that you need to use ``git-lfs`` to clone the repository.
- I was able to run the tests found in ``blimpy/tests/``, but the code coverage is currently only 44%. It would be good if this was a bit higher. (Note: If you have code that doesn't require testing, you can flag it using ``# pragma: no cover``. Then it won't be included in coverage calculations.)

Currently, I'm not sure what else I can do with ``Blimpy``. I might need to wait until there are more examples to get me started.

**Comment by telegraphic** [comment]
Thanks @garrettj403 (et al), we're having a look now and will get back to you soon!

**Comment by zhampel** [comment]
@telegraphic 

After installing via `pip install blimpy` and running `python setup.py test`, I receive an error regarding the lack of module `bitshuffle`. Any guidance on properly prepping & running the test suite?

Regarding the docs, I'm not sure the 'Writing Docs' page is necessary nor particularly instructive for the purposes of contributing to the blimpy codebase. I think a contributing guide similar to [this](https://jrbourbeau.github.io/pyunfold/contributing.html) is more like what this requirement is for JOSS publication. I also agree with @garrettj403 that the API is rather unclear.

I think until further example usage and more in depth documentation is provided, I can't provide further review of your submission.

**Comment by garrettj403** [comment]
@zhampel 

They have a couple extra packages listed under ``extras_require`` in ``setup.py``, including bitshuffle. From your local copy of blimpy, you can run: ``pip install -e .[full]``

Again, this should be explained in the installation instructions.

**Comment by xuanxu** [comment]
:wave: @telegraphic: Have you had the opportunity to address the issues raised by the reviewers?

**Comment by telegraphic** [comment]
Hey @xuanxu , we've had a meeting about moving this forward, and I'll be devoting some time to it this weekend. Quite a few suggestions!

**Comment by telegraphic** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

```

**Comment by telegraphic** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf)

**Comment by telegraphic** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf)

**Comment by telegraphic** [comment]
Hi @xuanxu, getting there: I'm waiting on a pull request to be reviewed/accepted that addresses a lot of the points raised. (One of the authors just finished their PhD so took some well-deserved time off)

**Comment by telegraphic** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1554 with the following error: 

 /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse': (tmp/1554/paper.md): mapping values are not allowed in this context at line 17 column 17 (Psych::SyntaxError)
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse_stream'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:325:in `parse'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:252:in `load'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:473:in `block in load_file'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `open'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `load_file'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/lib/whedon.rb:115:in `load_yaml'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/lib/whedon.rb:85:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-efe915e61673/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by telegraphic** [comment]
Hmm, whedon is not happy. I just removed a single space in the metadata of `paper.md`, let's see if that fixes it

**Comment by telegraphic** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf)

**Comment by telegraphic** [comment]
LGTM Whedon! 

@xuanxu, @garrettj403  and @zhampel, thanks for your patience. We have made quite a few changes thanks to your suggestions:

1) Firstly, we have updated the paper.md with references and an expanded author list. The README is also updated, and I've added a link to our [data format paper](https://ui.adsabs.harvard.edu/abs/2019arXiv190607391L/abstract), that is now accepted to PASP (Publications of the Astronomical Society of the Pacific). 

2) We removed the need for git-lfs, and now data are downloaded from our self-hosted data archive. We've also moved code coverage over from coveralls over to codecov.io, and worked on improving the coverage to 70%.

3) I added a [jupyter notebook example](https://github.com/UCBerkeleySETI/blimpy/blob/master/examples/voyager.ipynb) into the new `examples` directory, which shows how to use the code to find the telemetry signal from the Voyager telescope from some of our data.

4) We also made a structural change to the package -- you'll notice the code has been rearranged into directories, based on what the code does (as a sidenote, apart from being logical, this makes the coverage reports much more useful!)
  
BEFORE:

```
blimpy
  |- calib_utils
  |- deprecated
```

AFTER

```
blimpy
  |- calib_utils
  |- deprecated
  |- plotting
  |- ephemeris
  |- io
```

* The `io` directory contains `file_wrapper.py`, the main file handler, `fil_writer.py` and `hdf_writer.py` for writing files, and `sigproc.py` for handling sigproc filterbank files. Anyone wanting to add a new reader or writer need only touch code in this directory.

* The `ephemeris` directory contains `compute_lst.py` for computing sidereal time, and `compute_lsrk.py`, which computes the LSRK for doppler correction of Earth rotation. The lat/lon/elevation of the Parkes and GBT telescopes is included in `config.py` here, so that other telescopes can be added straightforwardly in the future.

* The `plotting` directory houses all the methods for plotting, that were previously methods of the Filterbank/Waterfall classes (they are attached to the class using a create_bound_method function).

**Comment by garrettj403** [comment]
Hi @telegraphic 

The recent changes look good. It's much easier to understand the functionality with the new changes to the paper and the workflow example.

I still have a few remaining comments that I would like to be addressed:
- I can't see an archive for Blimpy (i.e., through a 3rd party archiving service such as Zenodo)
- The version is 1.3.5 on GitHub, but 1.4.0 on PyPI
- The package API is still a bit of a mess. A lot of the functions are missing descriptions of what they do as well as descriptions of the function arguments.
- The package is missing community guidelines (i.e., "guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support")
- The automated tests currently aren't passing. On Travis, it says that they are passing, but it's not actually collecting any tests (i.e., ``collected 0 items / 1 errors``). It looks like [build 368](https://travis-ci.org/UCBerkeleySETI/blimpy/jobs/592371909) worked correctly. Also, running ``pytest tests/`` on my computer causes it to crash, but that could be because I don't have enough memory.

I'm happy to sign off once these last few issues are fixed (especially the API).

**Comment by xuanxu** [comment]
About the Zenodo archive: There is no need to archive the code before the review is finished. That way the archive created will include all changes made during the review process.

**Comment by telegraphic** [comment]
Hi @garrettj403 and @xuanxu, I've got a draft in Zenodo ready -- it looks like I can fill in the Journal ID as metadata, so I'll fill this in once the review is finished?

We've fixed the Travis-CI problem, added CONTRIBUTIONS.md, and I went through and improved docstrings. I've bumped to v 1.4.1 also, which we will push to PyPi (relevant collaborator is currently on vacation, so might take a few days).

**Comment by garrettj403** [comment]
Hi @telegraphic, the changes to the docstrings look good. For some reason [``blimpy.waterfall.Waterfall``](https://blimpy.readthedocs.io/en/latest/blimpy.html#blimpy.waterfall.Waterfall) doesn't display all of the arguments correctly online, but it does display them correctly using ``pydoc``.

@xuanxu, I'm ready to recommend publication. The documentation is very nice and it includes an [example workflow](https://github.com/UCBerkeleySETI/blimpy/blob/master/examples/voyager.ipynb) to get new users started. The package is easy to install via ``pip``. (It's a bit harder if you want to run the unit tests, but they have included instructions in their ``README`` to tell you how to do that.) Overall, it is an interesting package that allows you to interact with Breakthrough Listen data.

**Comment by xuanxu** [comment]
@garrettj403 Thanks!

**Comment by xuanxu** [comment]
:wave: @zhampel can you review the changes [made by @telegraphic](https://github.com/openjournals/joss-reviews/issues/1554#issuecomment-536267054) and summarize what points (if any) you feel are still remaining? Thanks!

**Comment by zhampel** [comment]
@xuanxu I will try to get to it by the end of the week. I have had a computer failure yesterday so I can’t reasonably re-review till then.

**Comment by zhampel** [comment]
@telegraphic 

Thanks for the reorganization of the codebase, as it's clearer how the pieces work. The example notebook is a nice addition for usage purposes. Furthermore, thanks for the clarifying the README for streamlined install, testing, usage, and some simple module calls.

@xuanxu I recommend this submission for publication.

**Comment by xuanxu** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf)

**Comment by xuanxu** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.3847/1538-4357/aa8d1b is OK
- 10.3847/2515-5172/aaa6c9 is OK
- 10.3847/2515-5172/ab010b is OK
- 10.1017/S1473550417000465 is OK
- 10.3847/1538-4357/aad005 is OK
- 10.1109/MCSE.2011.37 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by xuanxu** [comment]
@telegraphic: a couple of minor details to fix in the paper:
- from the summary of the paper: `Over 1 PB of from the...`. I guess there's a `data`missing there, so it should be `Over 1 PB of *data* from the...`
- at the end of the summary, when you mention the publications that have already used Blimpy, the `Enriquez et al. 2019` reference is duplicated

**Comment by telegraphic** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf)

**Comment by telegraphic** [comment]
@xuanxu missing 'data' added. I think the bib style is generating 'Price, Croft et al' and 'Price, Enriquez et al', i.e. showing first two authors, to differentiate Price2019a and Price2019b?

**Comment by xuanxu** [comment]
@whedon set 1.4.1 as version

**Comment by whedon** [comment]
OK. 1.4.1 is the version.

**Comment by xuanxu** [comment]
Thanks @telegraphic, everything looks good, here are the next steps:

- Please archive the latest release in Zenodo
- Check the Zenodo deposit has the correct metadata: title and authors names should match the paper; you may also add your ORCID.

Once you do that please report here the Zenodo DOI.

**Comment by telegraphic** [comment]
Done! 

10.5281/zenodo.3472084

**Comment by xuanxu** [comment]
@whedon set 10.5281/zenodo.3472084 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3472084" target="_blank">10.5281/zenodo.3472084</a> is the archive.

**Comment by xuanxu** [comment]
Everything looks ready 🎉
Pinging @openjournals/joss-eics for final acceptance

**Comment by Kevin-Mattheus-Moerman** [comment]
@telegraphic some minor points about the paper

- [ ] Check your affiliations. The second instance of `Nijmegen` is spelled wrong, it currently says `Nimegen`. Also I believe the correct affiliation there should probably be `Radboud University, Nijmegen, the Netherlands`. Please also add `United States of America`, and spell out `the Netherlands`. 
- [ ] In your repo and the Zenodo archive Blimpy is written with a capital B. The paper uses a lower case latter. Can you amend the paper to use a capital letter too? Alternatively change everything to have a lower case letter instead, whatever you prefer. 
- [ ] Give your paper a final check again yourself as we'll shortly process formal acceptance.

**Comment by telegraphic** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01554/joss.01554/10.21105.joss.01554.pdf)

**Comment by telegraphic** [comment]
@Kevin-Mattheus-Moerman Thanks, just made the suggested changes and regenerated the PDF.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.3847/1538-4357/aa8d1b is OK
- 10.3847/2515-5172/aaa6c9 is OK
- 10.3847/2515-5172/ab010b is OK
- 10.1017/S1473550417000465 is OK
- 10.3847/1538-4357/aad005 is OK
- 10.1109/MCSE.2011.37 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1037

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1037, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1186198675560783878) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1038
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01554](https://doi.org/10.21105/joss.01554)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01554/status.svg)](https://doi.org/10.21105/joss.01554)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01554">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01554/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01554/status.svg
   :target: https://doi.org/10.21105/joss.01554
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01554/status.svg)](https://doi.org/10.21105/joss.01554)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
